File: MCL_by_python/Points.py
from __future__ import print_function
import BeamModel
import ButtonCommand
import InitializeMap
import VelocityModel
import Map
import Robot
import math
import random

global ParticleCount, DeltaR, DeltaTheta, Particles, m_AggregatedWeights, m_TempNewParticles


def IsInside(x, y):
    isIn = False
    for i in range(len(Map.MapPoints)):
        j = (i + len(Map.MapPoints) - 1) % len(Map.MapPoints)
        if ((Map.MapPoints[i][1] > y) != (Map.MapPoints[j][1] > y)) and (
                x < (Map.MapPoints[j][0] - Map.MapPoints[i][0]) * (y - Map.MapPoints[i][1]) * 1.0 / (
                Map.MapPoints[j][1] - Map.MapPoints[i][1]) + Map.MapPoints[i][0]):
            isIn = not isIn
    return isIn


def CreateRandomPose():
    head = random.random() * (random.randint(0, 360))
    while (1):
        x = Map.XMin + random.random() * (Map.XMax - Map.XMin)
        y = Map.YMin + random.random() * (Map.YMax - Map.YMin)
        if IsInside(x, y):
            break
    return x, y, head


def InitializeParticles():
    global ParticleCount, Particles
    for i in range(ParticleCount):
        Particles[i] = CreateRandomPose()

        # if IsInside(Particles[i][0], Particles[i][1]):
        InitializeMap.InitPoints(i, Particles[i][0], Particles[i][1])
        # else:
        #     Particles[i] = CreateRandomPose()


def CalculateWeight(pose, measurements):
    """pose:x,y,degree"""
    weight = 1.00
    for i in range(0, 5):
        headDeg = pose[2] + Robot.SensorDirections[i]

        distance_to_wall = Map.GetClosestWallDistance(pose[0], pose[1], headDeg)

        Probability = BeamModel.Getprobability(measurements[i], distance_to_wall, DeltaR)
        weight = weight * Probability

    return weight


def DrawParticle(Random):
    global ParticleCount, Particles, m_AggregatedWeights
    # print('Points.61.Random = %d' % Random)
    low_index = 0
    high_index = ParticleCount - 1

    while low_index <= high_index:

        mid_num = int((low_index + high_index) / 2)
        # print('Points.61.mid_num = %d' % mid_num)

        range_lower = 0 if mid_num - 1 < 0 else m_AggregatedWeights[mid_num - 1]
        range_upper = m_AggregatedWeights[mid_num]

        if Random < range_lower:
            high_index = mid_num - 1
        elif Random >= range_upper:
            low_index = mid_num + 1
        elif range_lower <= Random < range_upper:
            return Particles[mid_num]
        else:
            print('Error')


def DrawParticles(sumOfWeights):
    global ParticleCount, m_TempNewParticles

    # global Particles
    m_swap = [[0] * 3 for i in range(ParticleCount)]
    for i in range(ParticleCount):
        Random = random.random() * sumOfWeights
        # DrawParticle(Random)
        m_TempNewParticles[i] = DrawParticle(Random)
        # InitializeMap.DrawPoints(m_TempNewParticles[i][0], m_TempNewParticles[i][1], m_TempNewParticles[i][2])
        m_swap = Particles[i]
        Particles[i] = m_TempNewParticles[i]
        m_TempNewParticles[i] = m_swap


def update(measurements):
    global ParticleCount, Particles, m_AggregatedWeights
    weights_sum = 0
    movedParticle = [0, 0, 0]  # X,Y,degree
    for i in range(ParticleCount):
        original_particle = Particles[i]
        movedParticle = VelocityModel.poseSample(original_particle, ButtonCommand.currentStep)

        while (not IsInside(movedParticle[0], movedParticle[1])):
            Particles[i] = CreateRandomPose()
            movedParticle = VelocityModel.poseSample(original_particle, ButtonCommand.currentStep)
            # if IsInside(Particles[i][0], Particles[i][1]):
            #     break

        # print movedParticle
        weight = CalculateWeight(movedParticle, measurements)

        weights_sum = weights_sum + weight
        m_AggregatedWeights[i] = weights_sum

        Particles[i] = movedParticle
    DrawParticles(weights_sum)


def MonteCarloLocalization(particleCount):
    global ParticleCount, DeltaR, DeltaTheta, Particles, m_AggregatedWeights, m_TempNewParticles

    ParticleCount = particleCount
    DeltaR = 3.0
    DeltaTheta = math.radians(5.0)

    Particles = [[0] * 3 for i in range(ParticleCount)]
    m_AggregatedWeights = [[0] for i in range(ParticleCount)]
    m_TempNewParticles = [[0] * 3 for i in range(ParticleCount)]

    InitializeParticles()


def updateUI():
    InitializeMap.delete()
    InitializeMap.DrawParticles()

File: MCL_by_Cpp/MCL_C++/Points.cpp
#include "stdafx.h"
#include <iostream>
#include "stdlib.h"
#include <cmath>
#include <ctime>
#include "string.h"
#include "Points.h"
#include "BeamModel.h"
#include "CommonFunctions.h"
#include "Drive.h"
#include "Map.h"
#include "Robot.h"
#include "VelocityModel.h"
using namespace std;

#define random(x) (rand()%x)

int ParticlesCount = 0;
double DeltaR = 0;
double m_AggregatedWeights[10000] = {0};
Pose m_TempNewParticles[10000] = {0};
Pose Particles[10000] = {0};

//Pose::Pose(double x, double y, double theta): x(x), y(y), theta(theta) {};

void Pose::printX()
{
    cout << "x = " << x << endl;
    cout << "y = " << y << endl;
    cout << "theta = " << theta << endl;
};

double Pose::rad()
{
  double temp = theta;
  double flag = (theta < 0) ? -1.0 : 1.0;  //  Judgement the positive or negative
  if (theta < 0)
    temp = theta * (-1.0);
  
  return   flag * (temp * p_PI) / 180;
};

// double Pose::ang(double angle)
// {
//   
//   double flag = (angle < 0) ? -1.0 : 1.0;  //  Judgement the positive or negative
//   if (angle < 0)
//     angle = angle * (-1.0);
//   
//   return flag * (angle * p_PI) / 180;
// 
// };


double rad2ang(double rad)
{
  double flag = (rad < 0) ? -1.0 : 1.0;  //  Judgment the positive or negative
  if (rad < 0)
    rad = rad * (-1.0);
  
  return flag * (rad * 180) / p_PI;
}

double ang2rad(double angle)
{
  double flag = (angle < 0) ? -1.0 : 1.0;  //  Judgment the positive or negative
  if (angle < 0)
    angle = angle * (-1.0);
  
 return flag * (angle * p_PI) / 180;
}



Pose CreateRandomPose()
{
  // srand((int)time(0));
  Pose CRPose = { 0 };
  CRPose.theta= random(1000)%360;
  do
  {
	  CRPose.x = XMin + GenerateRandom()*(XMax - XMin);
	  CRPose.y = YMin + GenerateRandom()*(YMax - YMin);
  } while (!IsInside(CRPose.x, CRPose.y));
  
  return CRPose;
}

void InitializeParticles()
{
  for (int i=0; i<ParticlesCount; i++)
  {
    Particles[i] = CreateRandomPose();
    InitParicle(Particles[i].x, Particles[i].y);
  }
}

void MonteCarloLocalization(int particleCount)
{
  ///  Initial parameter
  
  ParticlesCount = particleCount;
  //DeltaR = 3.0;
  
  InitializeParticles();
}


double CalculateWeight(Pose pose, double measurements[5])
{
    ///  pose:x,y,degree
  
    double weight = 1.00;
    double headAng, distanceToWall, Probability;
    for (int i=0; i<SenLen;i++)
    {
      headAng = pose.theta + Sensor[i];
      distanceToWall = GetClosestWallDistance(pose.x, pose.y, headAng);

	  weight *= Getprobability(measurements[i], distanceToWall, DeltaR);
      
    }
    return weight;
}


Pose DrawParticle(double random)
{
	// binary search for the index; see http://www.dreamincode.net/code/snippet1835.htm
	// 二进制搜索索引
	int lowIndex = 0;
	int highIndex = ParticlesCount;

	// loop while the low index is less or equal to the high index
	// 当低指数小于或等于高指数时循环
	while (lowIndex <= highIndex)
	{
		//get the middle point in the array
		int midNum = (lowIndex + highIndex) / 2;

		double rangeLower = midNum - 1 < 0 ? 0 : m_AggregatedWeights[midNum - 1];
		double rangeUpper = m_AggregatedWeights[midNum];

		//now start checking the values
		if (random < rangeLower)
		{
			//search value is lower than this index of our array
			//so set the high number equal to the middle number - 1
			highIndex = midNum - 1;
		}
		else if (random >= rangeUpper)
		{
			//search value is higher than this index of our array
			//so set the low number to the middle number + 1
			lowIndex = midNum + 1;
		}
		else if (random >= rangeLower && random < rangeUpper)
		{
			//we found a match
			//Debug.WriteLine(rangeLower + " - " + rangeUpper + ": " + random);
			return Particles[midNum];
		}
	}
	cout << "Program error" << endl;
    return ;
}


void DrawParticles(double sumOfWeights)
{
  /// Redistribute particles
  
  Pose m_swap[10000] = {0};
  double Random;
  for(int i=0; i<ParticlesCount; i++)
  {
        Random = GenerateRandom() * sumOfWeights;
        //DrawParticle(Random)
        m_TempNewParticles[i] = DrawParticle(Random);
        //InitializeMap.DrawPoints(m_TempNewParticles[i][0], m_TempNewParticles[i][1], m_TempNewParticles[i][2]);    
  }
  memcpy(m_swap, Particles, sizeof(Particles));
  memcpy(Particles, m_TempNewParticles, sizeof(Particles));
  memcpy(m_TempNewParticles, m_swap, sizeof(m_AggregatedWeights));
}

void update(double measurements[5])
{
  /* Calculate each particle weight, 
     calculate the weight accumulation sum, 
      and update the particle distribution. */

  double weights_sum = 0;
  Pose MovedParticle = { 0, 0, 0 };  // x, y, angle
  Pose OriParticle = { 0, 0, 0 };  // x, y, angle
    
  for (int i=0; i<ParticlesCount; i++)
  {
	  //cout << i << endl;
	  if (i == 500)
		  i = i;
    OriParticle = Particles[i];
    MovedParticle = poseSample(OriParticle);
      
    while (!IsInside(MovedParticle.x, MovedParticle.y))
    {
      OriParticle = CreateRandomPose();
      MovedParticle = poseSample(OriParticle);
    }

    double weight = CalculateWeight(MovedParticle, measurements);

    weights_sum = weights_sum + weight;
    m_AggregatedWeights[i] = weights_sum;
      
    Particles[i] = MovedParticle; // 
  }
    
  DrawParticles(weights_sum);
}